hack/generator/pkg/astmodel/type_walker.go

```diff
@@ -0,0 +1,195 @@
+	processing map[TypeName]struct{}
+}
+
+// NewTypeWalker returns a TypeWalker.
+// The provided visitor VisitTypeName function must return a TypeName and VisitObjectType must return an ObjectType or calls to
+// Walk will panic.
+func NewTypeWalker(allTypes Types, visitor TypeVisitor) *TypeWalker {
+	typeWalker := TypeWalker{
+		allTypes: allTypes,
+	}
+	typeWalker.originalVisitTypeName = visitor.VisitTypeName
+	typeWalker.originalVisitObjectType = visitor.VisitObjectType
+
+	// visitor is a copy - modifications won't impact passed visitor
+	visitor.VisitTypeName = typeWalker.visitTypeName
+	visitor.VisitObjectType = typeWalker.visitObjectType
+
+	typeWalker.visitor = visitor
+	typeWalker.AfterVisitFunc = DefaultAfterVisit
+	typeWalker.MakeContextFunc = IdentityMakeContext
+	typeWalker.WalkCycle = IdentityWalkCycle
+
+	return &typeWalker
+}
+
+func (t *TypeWalker) visitTypeName(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error) {
+	updatedCtx, err := t.MakeContextFunc(it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	visitedTypeName, err := t.originalVisitTypeName(this, it, updatedCtx)
+	if err != nil {
+		return nil, err
+	}
+	var ok bool
+	it, ok = visitedTypeName.(TypeName)
+	if !ok {
+		panic(fmt.Sprintf("TypeWalker visitor VisitTypeName must return a TypeName, instead returned %T", visitedTypeName))
+	}
+
+	def, ok := t.allTypes[it]
+	if !ok {
+		return nil, errors.Errorf("couldn't find type %q", it)
+	}
+
+	// Prevent loops by bypassing this type if it's currently being processed. The processing
+	// slice is basically the "path" taken to get to the current type.
+	if _, ok := t.state.processing[def.Name()]; ok {
+		return t.WalkCycle(def, updatedCtx)
+	}
+	t.state.processing[def.Name()] = struct{}{}
+
+	updatedType, err := this.Visit(def.Type(), updatedCtx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error visiting type %q", def.Name())
+	}
+	updatedDef, err := t.AfterVisitFunc(def, def.WithType(updatedType), updatedCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	delete(t.state.processing, def.Name())
+
+	err = t.state.result.AddWithEqualityCheck(updatedDef)
+	if err != nil {
+		return nil, err
+	}
+
+	return updatedDef.Name(), nil
+}
+
+func (t *TypeWalker) visitObjectType(this *TypeVisitor, it *ObjectType, ctx interface{}) (Type, error) {
+	result, err := t.originalVisitObjectType(this, it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ot, ok := result.(*ObjectType)
+	if !ok {
+		panic(fmt.Sprintf("TypeWalker visitor VisitObjectType must return a *ObjectType, instead returned %T", result))
+	}
+
+	for _, prop := range ot.Properties() {
+		isCycle := isRemoveType(prop.PropertyType())
+		if isCycle {
+			ot = ot.WithoutProperty(prop.PropertyName())
+		}
+	}
+
+	return ot, nil
+}
+
+// Walk returns a Types collection constructed by applying the Visitor to each type in the graph of types reachable
+// from the provided TypeDefinition 'def'. Types are visited in a depth-first order. Cycles are not visited.
```

Comment:
Is it right that cycles aren't visited? I think it's that the walk down the graph will stop when it detects a cycle - so the types in the cycle will be visited but only once. But I could definitely be wrong.

Reply:
No you're exactly right. I've corrected this language to:
> Types are visited in a depth-first order. Cycles are not followed (so each type in a cycle will be visited only once).